File: src/main/java/fi/quanfoxes/Chain.java
package fi.quanfoxes;

import java.util.concurrent.ExecutorService;

public class Chain {
    private ExecutorService executors;
    private Class<? extends Phase>[] phases;

    @SafeVarargs
    public Chain(ExecutorService executors, Class<? extends Phase>... phases) {
        this.executors = executors;
        this.phases = phases;
    }

    public void execute(Bundle bundle) {
        for (var template : phases) {

            boolean multithreaded = bundle.getBool("multithreaded", true);

            try {
                Phase phase = template.getConstructor().newInstance();
                phase.setExecutors(executors, multithreaded);
                phase.execute(bundle);
                phase.sync();

                if (!phase.succeeded()) {
                    break;
                }
            }    
            catch (Exception e) {
                System.err.println("Internal error: " + e.getMessage());
                return;
            }
        }
    }
}

File: Zigzag/Chain.cs
using System;

public class Chain
{
    private System.Type[] Phases { get; set; }

    /// <summary>
    /// Creates a chain of phases
    /// </summary>
    /// <param name="phases">Phases to execute in order when invoked</param>
    public Chain(params System.Type[] phases)
    {
        Phases = phases;
    }

    /// <summary>
    /// Executes the configured phases with the given bundle
    /// </summary>
    /// <param name="bundle">Bundle to pass to the phases</param>
    public bool Execute(Bundle bundle)
    {
        foreach (var template in Phases)
        {
            var multithreaded = bundle.Get("multithreaded", true);

            try
            {
                if (!(Activator.CreateInstance(template) is Phase phase))
                {
                    throw new ApplicationException("Couldn't create the next phase");
                }

                phase.Multithread = multithreaded;

                var status = phase.Execute(bundle);

                if (status.IsProblematic)
                {
                    Console.Error.WriteLine($"Terminated: {status.Description}");
                    return false;
                }

                phase.Sync();

                if (status.IsProblematic)
                {
                    Console.Error.WriteLine($"Terminated: {status.Description}");
                    return false;
                }
                else if (phase.Failed)
                {
                    Console.Error.WriteLine($"Terminated: {phase.GetTaskErrors()}");
                    return false;
                }
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("Internal error: " + e.Message);
                return false;
            }
        }


		return true;
	}
}